### PR TITLE
fix elliptic signing

### LINKED ISSUE #9237: Fix ODIS authentication

### PR DESCRIPTION
**Pull request by alecps**
### Other changes

Removes elliptic as a dependency from ODIS signer and combiner 

Updates get-contact-matches to rely on the common package's implementation of verifyDEKSignature

[adds flake tracker to @celo/keystore and increases jest.timeout](https://github.com/celo-org/celo-monorepo/pull/9271/commits/2112edc37c0daf7171103e872cd7d39199d35d2f)

### Tested

Test cases added to phone-number-privacy/common package ensuring that authenticateUser accepts both the new and the old signing schemes (backwards compatibility)

Test cases added to @celo/identity ensuring that signWithRawKey signs the full message digest rather than the plain text message.

### Related issues

- Fixes #9237 

### Backwards compatibility

The verifyDEKSignature function used by authenticateUser() will now accept requests with both the old (incorrect) and the new (correct) authorization signatures. Once clients have upgraded to use this change in the @celo/identity package, the verifyDEKSignature function will be updated again to only accept the new authorization signatures. 

To monitor frequency of requests with incorrect signatures, we have added a new WarningMessage so that we can easily create a log based metric. We will not remove support for the old signing method until ODIS is no longer receiving requests that use it. 

### Documentation

None

### CODE REVIEW

File: packages/phone-number-privacy/common/src/interfaces/error-utils.ts [hunk 1]
```diff
@@ -30,4 +30,5 @@ export enum WarningMessage {
   CANCELLED_REQUEST_TO_SIGNER = 'CELO_ODIS_WARN_09 SIGNER Cancelled request to signer',
   INVALID_USER_PHONE_NUMBER_SIGNATURE = 'CELO_ODIS_WARN_10 BAD_INPUT User phone number signature is invalid',
   UNKNOWN_DOMAIN = 'CELO_ODIS_WARN_11 BAD_INPUT Provided domain name and version is not recognized',
+  INVALID_AUTH_SIGNATURE = 'CELO_ODIS_WARN_12 BAD_INPUT Authorization signature was incorrectly generated',
```
Comment (nategraf) [inline packages/phone-number-privacy/common/src/interfaces/error-utils.ts]: Seeing this warning, I would have assumed the request would be rejected. I added a suggestion below to indicate that it is not, but will be in the future.
```suggestion
  INVALID_AUTH_SIGNATURE = 'CELO_ODIS_WARN_12 BAD_INPUT Authorization signature was incorrectly generated. Request will be rejected in a future version.',
```

File: packages/phone-number-privacy/common/src/utils/authentication.ts [hunk 2]
```diff
@@ -66,16 +68,26 @@ export function verifyDEKSignature(
   registeredEncryptionKey: string,
   logger?: Logger
 ) {
+  logger = logger ?? rootLogger()
   try {
+    const msgDigest = crypto.createHash('sha256').update(JSON.stringify(message)).digest('base64')
```
Comment (nategraf) [inline packages/phone-number-privacy/common/src/utils/authentication.ts]: If I read the `elliptic` docs correctly, the message digest needs to be `hex` encoded. In fact, if I read the code correctly, this will lead to incorrect behavior where it interprets the `base64` as `hex` with strange results. Especially when dealing with a library with such a terrible interface, we need tests that can catch this kind of error.

File: packages/phone-number-privacy/common/src/utils/authentication.ts [hunk 3]
```diff
@@ -66,16 +68,26 @@ export function verifyDEKSignature(
...
-  } catch (err) {
-    if (logger) {
-      logger.error('Failed to verify signature with DEK')
-      logger.error({ err, dek: registeredEncryptionKey })
+    if (key.verify(msgDigest, parsedSig)) {
+      return true
+    }
+
+    if (key.verify(message, parsedSig)) {
+      // TODO: Remove this once clients upgrade to @celo/identity v1.5.3
```
Comment (nategraf) [inline packages/phone-number-privacy/common/src/utils/authentication.ts]: We should be more explicit about why this code path exists and why it needs to be removed. Also, we should not write a `verify` function that exhibits a strange behavior like this without sufficient warning to the user. I suggest we add an (intentionally ugly) `insecureAllowIncorrectlyGeneratedSignature` parameter that should be set to `true` in order to activate this code path.
```suggestion
    // TODO: Remove this once clients upgrade to @celo/identity v1.5.3
    // Due to an error in the original implementation of the sign and verify functions
    // used here, older clients may generate signatures over the truncated message,
    // instead of its hash. These signatures represent a risk to the signer as they do
    // not protect against modifications of the message past the first 64 characters of the message.
    if (insecureAllowIncorrectlyGeneratedSignature && key.verify(message, parsedSig)) {
```

File: packages/sdk/identity/src/odis/query.test.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import crypto from 'crypto'
+import { hexToBuffer, trimLeading0x } from '../../../base/lib'
+import { signWithRawKey } from './query'
+
+const rawKey = '41e8e8593108eeedcbded883b8af34d2f028710355c57f4c10a056b72486aa04'
+
+describe(signWithRawKey, () => {
+  it('Signs full message digest', async () => {
+    const msg = 'test'
```
Comment (nategraf) [inline packages/sdk/identity/src/odis/query.test.ts]: This message is too short to actually test the vulnerability. It needs to be more than 64 characters. The main vulnerability is due to message truncation. We should have a test that verifies that changes to the message invalidates the signature.

File: packages/sdk/identity/tslint.json [hunk 5]
```diff
@@ -4,6 +4,7 @@
     "no-global-arrow-functions": false,
     "no-console": false,
     "member-ordering": false,
-    "max-classes-per-file": false
+    "max-classes-per-file": false,
+    "import-blacklist": [true, { "elliptic": ["ec"]}]
```
Comment (nategraf) [inline packages/sdk/identity/tslint.json:8]: Nice thinking! Thanks for putting this library on the black-list. It definitly hazmat.

**Review by nategraf — COMMENTED** [review]
Major comments are:
* I believe the encoding of the message digest needs to be `hex`. IIUC, using `base64` is incorrect and may actually lead to vulnerable behavior.
* We need more testing to confirm both positive and negative test cases (e.g. that the signature verify when we expect and _don't_ when we expect them not to). Examples include testing that modifying the message anywhere, modifying signature, modifying the key, using a different key, of using an inconsistent authentication mode all result in a verification error.
* If at all possible, we should add tests to verify cross-compatibility of this `elliptic` implementation with another standard implementation. It looks like the `crypto` library supports `secp256k1` ECDSA signatures on systems with the `openssl` library support. We should try adding a test that signs with our `elliptic` based signing function and verified with `crypto` and vice versa (this would catch the `base64` error, unless it actually isn't an error, and would have caught the original issue.)
* Along with a release, we should publish a CVE for the client library. Hopefully the security team can help us with that. I believe we know everywhere this code is used, but we need to make cure any client using it updates in a timely fashion. Along with that we should have an issue to track the need to monitor uptake and disable the permissive verification. We should consider setting a deadline for clients to upgrade.

**Review by codyborn — APPROVED** [review]
Looks good! Thanks for the quick fix.

**Review by nategraf — APPROVED** [review]
In this PR https://github.com/celo-org/celo-monorepo/pull/9290, I added some extra tests, as well as removing an unnecessary verification during signing. With that, I think this is good to go.

From now on, I would consider `elliptic` to be hazmat. Definitly glad it is on the blacklist. If we ever need signature authentication again, we should work to avoid implementing it at such a low level.
